Epic/49688 custom set of project attributes grouped in sections

jjabari-op opened the pull request:
https://community.openproject.org/projects/stadt-koeln/work_packages/49688

Includes:
 - https://community.openproject.org/wp/53007

jjabari-op committed:
[#49688] Individuelle Gruppen für Projektattribute

https://community.openproject.org/work_packages/49688

jjabari-op committed:
WIP: first steps introducing project attributes on project overview page, primerized project attribute editing via dialog

jjabari-op committed:
introduce project custom field sections and project mapping

jjabari-op committed:
introducing project attributes management with generic drag and drop approach

jjabari-op committed:
introducing project custom field create and edit

jjabari-op committed:
refactored datamodeling, reuse custom field forms in new project attribute context

jjabari-op committed:
fixed active menu item

jjabari-op committed:
disable project custom field management through custom fields UI

jjabari-op committed:
refactored project attributes sidebar and dialog following previous refactorings

jjabari-op committed:
introduce project custom field mapping UI

jjabari-op committed:
disable auto enabling of new project custom fields

jjabari-op committed:
minor cleanup

jjabari-op committed:
Merge branch 'dev' into epic/49688-individuelle-gruppen-für-projektattribute

jjabari-op committed:
quick fix issue when no custom field is enabled for a project

jjabari-op committed:
setup auto-scroller for project custom field admin page as suggested by @oliverguenther

jjabari-op committed:
added client side project custom field mapping filter via stimulus

jjabari-op committed:
added missing format display to project custom field mapping

jjabari-op committed:
temporary fix of overview page width issue seen in firefox

jjabari-op committed:
fixed responsive behavior and non-sidebar support

jjabari-op committed:
enable reverting of elements if they are dropped outside of a valid target

jjabari-op committed:
refined empty section behavior as requested

jjabari-op committed:
edit project attribute via click on name in row

jjabari-op committed:
fixed translation for section edit action

jjabari-op committed:
Merge branch 'dev' into epic/49688-individuelle-gruppen-für-projektattribute

jjabari-op committed:
add bulk actions to project custom field mapping UI

jjabari-op committed:
added project count on global project attributes settings page

jjabari-op committed:
minor refactoring

jjabari-op committed:
switched to whole button instead of icon_button for toggling project custom field mappings

jjabari-op committed:
enable status code usage when responding with turbo streams

jjabari-op committed:
refactored project attributes update via project overview page dialogs and enable reset of multi value custom fields

jjabari-op committed:
fixed single user select list input scope

jjabari-op committed:
WIP: fixing user and multi value inputs

jjabari-op committed:
fixed redirect issue when deleting custom options

jjabari-op committed:
reworked custom field form due to renderig errors when using rails partials within a view_component

ulferts committed:
Merge remote-tracking branch 'origin/dev' into epic/49688-individuelle-gruppen-für-projektattribute

ulferts committed:
increase pp instance size

ulferts committed:
Merge remote-tracking branch 'origin/dev' into epic/49688-individuelle-gruppen-für-projektattribute

ulferts committed:
remove file with commented out code tripping up Zeitwerk

jjabari-op committed:
added feature flag and fixed minor bugs

jjabari-op committed:
introducing capybara specs for overview and project settings page

jjabari-op committed:
Merge branch 'dev' into epic/49688-individuelle-gruppen-für-projektattribute

jjabari-op committed:
updated Gemfile.lock

jjabari-op committed:
enabling project attributes feature for pullpreview env

jjabari-op committed:
adding more specs for project custom fields on overview page

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
Merge branch 'epic/49688-individuelle-gruppen-für-projektattribute' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
added specs for project custom field edit dialog, refactored form input implementation, fixed minor bugs

jjabari-op committed:
hiding more parts of the new concept behind the feature flag

jjabari-op committed:
refactored form implementations

jjabari-op committed:
disable editing for project members without editing permission

jjabari-op committed:
started to add validation specs

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

oliverguenther committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

oliverguenther committed:
Linting drag & drop controller

jjabari-op committed:
WIP: refactoring of prototypical code as discussed

jjabari-op committed:
fixed resetting of list fields

jjabari-op committed:
WIP: generalize custom field inputs a bit more, support single custom field form, minor refactoring

jjabari-op committed:
fixed form input name nesting

jjabari-op committed:
fixed dialog specs and boolean input

jjabari-op committed:
support rich text input

jjabari-op committed:
fixed dialog validation specs

jjabari-op committed:
refactored custom value update to contract/service based approach

jjabari-op committed:
add permission checks for overview page requests and added related specs

jjabari-op committed:
refactoring and name fixes

jjabari-op committed:
added basic permission spec for project custom field mapping UI

jjabari-op committed:
refactored project custom field mapping controller

jjabari-op committed:
include the acts_as_customizable patch only if project attributes feature is enabled

jjabari-op committed:
switched to primer's toggle component on mapping UI

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
adjust toggle switch based on feedback

jjabari-op committed:
cleanup

jjabari-op committed:
switch to op_primer header component

jjabari-op committed:
WIP: enable required custom fields automatically in all projects and disallow disabling them on project level

jjabari-op committed:
force all required project custom fields to be enabled in all projects

jjabari-op committed:
adjusted toggle switch position according to feedback

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
disable custom fields with default values when user explicitly provided a blank value

jjabari-op committed:
added project creation feature specs around project custom fields and their mapping

jjabari-op committed:
added project creation and copy feature specs around project custom fields and their activation

jjabari-op committed:
fixed default value display approach

jjabari-op committed:
added model specs for project custom field model hooks

jjabari-op committed:
preview long text values truncated and expand via dialog

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
show users with avatar component in sidebar

oliverguenther committed:
Set appendTo=body to fix position

jjabari-op committed:
quick fixing multi user submission

jjabari-op committed:
added dialog update specs and split spec file for better oversight

oliverguenther committed:
Fix selection of principal users

jjabari-op committed:
fixed icon only async dialog trigger button paddings

jjabari-op committed:
adjusted switch positions as requested

jjabari-op committed:
removed feature flag as discussed with @oliverguenther

jjabari-op committed:
fixed breadcrumb navigation

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
fixed sidebar display condition

jjabari-op committed:
switched to op header component

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

oliverguenther committed:
Base values on input so it's easier to pass custom options

oliverguenther committed:
Allow loading multiple resources

oliverguenther committed:
Fix multiple values being joined in a single input

oliverguenther committed:
Fix edit label

oliverguenther committed:
Remove user_autocompleter input in favor of just the one autocompleter

allows specifying which component to use, the inputs were almost the same anyway

jjabari-op committed:
fixed decorated select fields visibility and user select field multi value submission and clearing

jjabari-op committed:
fixed specs due to primer updates

jjabari-op committed:
fixing multi turbo-frame tags after stream update discovered while debugging specs

jjabari-op committed:
hide toggle switch loading indicator as requested by product team

jjabari-op committed:
adjusted descriptions as requested

jjabari-op committed:
use op header component

oliverguenther committed:
Fix detection of selected items

oliverguenther committed:
Fix min-height of autocompleter

jjabari-op committed:
added specs for project custom field administration, fixed minor bugs

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
adjusting project model customizable specs

jjabari-op committed:
added more project model customizable specs

jjabari-op committed:
fixed naming

jjabari-op committed:
fixed and enriched project csv export specs

jjabari-op committed:
fixed existing project custom field settings page specs

jjabari-op committed:
support easy usage of update method with implicit custom field activation

jjabari-op committed:
support implicit activation of custom fields when using custom_field_values method

jjabari-op committed:
fixed project index specs

jjabari-op committed:
fixing custom field scope after validation

jjabari-op committed:
addings spec for after validation state of dialog

jjabari-op committed:
added missing validation specs

jjabari-op committed:
adding more validation specs and fixing selected values state after validation for multi select fields

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
fixing project API

jjabari-op committed:
fixed not desired reactivation of custom fields which have been disabled in the past

dombesz committed:
Make the project attributes rich text display "Expand" link inline.

dombesz committed:
Format code

dombesz committed:
Fix validation messages not being displayed when a previous section was closed without saving.

dombesz committed:
Respect allow_non_open_versions setting on the version input, do not wait on non existence expectations.

dombesz committed:
Filter out projects with disabled custom fields on the projects index page

https://community.openproject.org/wp/53007

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
fixes XLS export specs

jjabari-op committed:
fixed custom value specs

jjabari-op committed:
fixing unit tests, mark last failing one as pending

dombesz committed:
Rename "Custom field section" to "Section" on attribute form

dombesz committed:
Replace the Project details widget with deprecation message

jjabari-op committed:
fix empty project attributes issue seen in specs

jjabari-op committed:
fixed and enriched project settings page specs

jjabari-op committed:
fixed dialog specs

dombesz commented on .github/workflows/pullpreview.yml:
```diff
@@ -45,7 +45,7 @@ jobs:
         admins: crohr,HDinger,machisuji,oliverguenther,ulferts,wielinde,cbliard
         compose_files: docker-compose.pullpreview.yml
-        instance_type: large
+        instance_type: xlarge
```
Is this still required?

jjabari-op replied:
Good question! @ulferts increased the size, but I don't know if it was just meant for this pullpreview instance...

dombesz commented on spec/models/custom_value_spec.rb:
```diff
@@ -150,7 +150,7 @@
 
     context "for a string custom field without default value" do
-      shared_let(:custom_field) { create(:project_custom_field, :string) }
+      shared_let(:custom_field) { create(:project_custom_field, :string, projects: [project]) }
```
I know this is not strictly related to the PR, but `shared_let` is supposed to have 1 definition in a spec file and it wasn't designed to be redefined, because the variable should stay the same during running the spec file. Redefining it this way, may also cause unexpected behaviour.

dombesz commented on db/migrate/20231031133334_create_project_custom_field_project_mappings.rb:
```diff
@@ -0,0 +1,34 @@
+class CreateProjectCustomFieldProjectMappings < ActiveRecord::Migration[7.0]
+  def up
+    create_table :project_custom_field_project_mappings do |t|
```
I think the name `project_custom_fields_projects` name would follow the rails convention of naming tables better.

jjabari-op replied:
True, but I do like the explicitness of `project_custom_field_project_mappings`. @oliverguenther any preference?

dombesz commented on app/models/project_custom_field_project_mapping.rb:
```diff
@@ -0,0 +1,45 @@
+                                    inverse_of: :project_custom_field_project_mappings
+
+  # # Additionally to the database-level unique constraint, the application-level validation ensures that a
+  # # custom_field is associated with only one section
```
At the moment, it is possible to save duplicate associations. Do we still need the uniqueness validator?

jjabari-op replied:
I've added a proper validation!

dombesz commented on app/contracts/project_custom_field_project_mappings/base_contract.rb:
```diff
@@ -0,0 +1,51 @@
+    validate :validate_has_select_project_custom_fields_permission
+    validate :validate_is_not_required
+
+    def validate_has_select_project_custom_fields_permission
```
I think having `validate_` in the method name is redundant. Wdyt think of the following name?
```suggestion
    def user_allowed_to_select_project_custom_fields
```

dombesz commented on app/contracts/project_custom_field_project_mappings/base_contract.rb:
```diff
@@ -0,0 +1,51 @@
+      errors.add :base, :error_unauthorized
+    end
+
+    def validate_is_not_required
```
```suggestion
    def is_not_required
```

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+    end
+
+    def sorted_available_custom_fields_by_section(section)
+      available_custom_fields_by_section(section)
```
This is the only place we use the `available_custom_fields_by_section`. Both methods are just 1 liners, I would rather merge them together.
```suggestion
      available_custom_fields
        .where(custom_field_section_id: section.id)
```

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+      # - boolean custom fields are always activated as a nil value is never provided (always true/false)
+      custom_field_ids = project.custom_values
+        .reject { |cv| cv.value.blank? }
+        .reject { |cv| cv.persisted? } # do not reactivate custom fields which have already been used and disabled
```
These two could be made in the same interation:
```suggestion
        .reject { |cv| cv.persisted? || cv.value.blank? } # do not reactivate custom fields which have already been used and disabled
```

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+      project_custom_field_project_mappings.where(custom_field_id: custom_field_ids).destroy_all
+    end
+
+    def query_available_custom_fields_on_global_level
```
I would rename this method to `def with_all_available_custom_fields`.

dombesz commented on app/components/settings/project_custom_fields/new_form_header_component.rb:
```diff
@@ -0,0 +1,40 @@
+      include ApplicationHelper
+      include OpPrimer::ComponentHelpers
+
+      def initialize
```
This seems to be unused.

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+      set_custom_field_values_method_from_acts_as_customizable_module(values)
+    end
+
+    # we cannot call super as the code in acts_as_customizable is shipped in a module
```
We can remove this method.

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+      # in order to support implicit activation of custom fields when values are provided during an update
+      self._query_available_custom_fields_on_global_level = true # set to false in after_save hook
+      set_custom_field_values_method_from_acts_as_customizable_module(values)
+    end
```
As I tested it, `super` seems to work, combining it with the renamed `with_all_available_custom_fields` this method could be simplified:

```suggestion
    # we need to query the available custom fields on a global level when updating custom field values
    # in order to support implicit activation of custom fields when values are provided during an update
    def custom_field_values=(values)
      with_all_available_custom_fields { super }
    end
```

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+          yield custom_field
+        end
+      end
+    end
```
This method also can be simplified by using the `with_all_available_custom_fields` and calling `super`.

```suggestion
    def for_custom_field_accessor(method_symbol)
      with_all_available_custom_fields { super }
    end
```

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+      else
+        true # validate all custom values if no specific section was specified
+      end
+    end
```
I think these 3 methods can be simplified:

```suggestion
    def validate_custom_values
      # validate custom values only of a specified section
      # instead of validating ALL custom values like done in acts_as_customizable
      custom_field_section_ids = CustomField
        .where(id: custom_field_values.pluck(:custom_field_id))
        .where(custom_field_section_id: _limit_custom_fields_validation_to_section_id)
        .pluck(:id)

      super(custom_field_section_ids)
    end
```

Then we can define a new optional parameter for the `validate_custom_values` in the `lib_static/plugins/acts_as_customizable/lib/acts_as_customizable.rb`

```ruby
def validate_custom_values(custom_field_ids=[])
  set_default_values! if new_record?

   custom_field_values
    .reject(&:marked_for_destruction?)
    .select { |cv| custom_field_ids.empty? || custom_field_ids.include?(cv.custom_field_id) }
    .select(&:invalid?)
    .each { |custom_value| add_custom_value_errors! custom_value }
end
```
This way we can provide a set of custom_field_ids that needs to be validated, instead of validating all the custom values.
Additionally, the `custom_field_section_ids` did not have any caching, so it was doing N+1 queries. With my suggestion the N+1 queries are also eliminated.

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+      end
+
+      custom_fields
+    end
```
```suggestion
    def available_custom_fields(global: false)
      # overrides acts_as_customizable
      # in contrast to acts_as_customizable, custom_fields are enabled per project
      # thus we need to check the project_custom_field_project_mappings
      custom_fields = ProjectCustomField
        .includes(:project_custom_field_section)

      # available_custom_fields is called from within the acts_as_customizable module
      # we don't want to adjust these calls, but need a way to query the available custom fields on a global level in some cases
      # thus we pass in this parameter as an instance flag implicitly here,
      # which is not nice but helps us to touch acts_as_customizable as little as possible
      #
      # additionally we provide the `global` parameter to allow querying the available custom fields on a global level
      # when we have explicit control over the call of `available_custom_fields`
      unless global || new_record? || _query_available_custom_fields_on_global_level
        custom_fields = custom_fields
          .where(id: project_custom_field_project_mappings.select(:custom_field_id))
          .or(ProjectCustomField.required)
      end

      custom_fields
    end
```

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+      self._query_available_custom_fields_on_global_level = nil
+    end
+
+    def reject_section_scoped_validation_for_creation
```
Perhaps the `Projects::CreateContract` would be a better place for this validation, because this error seems to be tied to the representation of the project custom fields, but not their internal workings.

jjabari-op replied:
I do agree partially, but this is strongly bound to the patching code in this file what makes me want to have it right there as well. Additionally, creating a project through the create contract is technically just one way of doing it. Using the project active record API standalone would then not trigger this important "validation". If you're ok with it, I tend to prefer having it bound to the model directly.

dombesz committed:
Refactoring on the project act_as_customizable_patch

dombesz commented on app/services/project_custom_field_project_mappings/bulk_edit_service.rb:
```diff
@@ -0,0 +1,103 @@
+#++
+
+module ProjectCustomFieldProjectMappings
+  class BulkEditService < ::BaseServices::BaseCallable
```
To follow the convention of services this should rather be called `BulkUpdateService`

dombesz commented on app/controllers/projects/settings/project_custom_fields_controller.rb:
```diff
@@ -0,0 +1,118 @@
+
+      update_sections_via_turbo_stream # update all sections in order not to mess with stimulus target references
+    else
+      # TODO: handle error
```
There are some todo comments left.

dombesz commented on app/controllers/projects/settings/project_custom_fields_controller.rb:
```diff
@@ -0,0 +1,118 @@
+  before_action :eager_load_project_custom_field_sections, only: %i[show toggle enable_all_of_section disable_all_of_section]
+  before_action :eager_load_project_custom_fields, only: %i[show toggle enable_all_of_section disable_all_of_section]
+  before_action :eager_load_project_custom_field_project_mappings,
+                only: %i[show toggle enable_all_of_section disable_all_of_section]
```
It seems like we could move the 3 before actions into one.

dombesz commented on app/controllers/projects/settings/project_custom_fields_controller.rb:
```diff
@@ -0,0 +1,118 @@
+  private
+
+  def eager_load_project_custom_field_sections
+    @project_custom_field_sections = ProjectCustomFieldSection.all.to_a
```
Is there a reason we need to explicitly call `to_a`? ActiveRecord seems to be loading a query only once.

jjabari-op replied:
From my understanding it is required in order to force the data to be fetched from the database and to be loaded in memory. I'm doing this in order to avoid N+1 query in view components afterwards.

dombesz replied:
I believe that will not prevent any N+1 queries, because when AR loads the results, it will not query the database each time the results are accessed.
However, we should definitely eager load the `custom_fields` association on the query, because it is referenced in the [project_custom_field_sections/show_component](https://github.com/opf/openproject/pull/14525/files#diff-ab7773fd1ebcdc5a97df0b4a20c6afe221b66a9be1c15d3ed51b0e2df5cfc0d5R40), which will indeed cause an N+1.

```suggestion
    @project_custom_field_sections = ProjectCustomFieldSection.includes(:custom_fields)
                                                         .order("custom_fields.position_in_custom_field_section ASC")
                                                         .all
```

Note: in order to not accidentally trigger the N+1 we should remove the `reorder(position_in_custom_field_section: :asc)` from [this line](https://github.com/opf/openproject/pull/14525/files#diff-ab7773fd1ebcdc5a97df0b4a20c6afe221b66a9be1c15d3ed51b0e2df5cfc0d5R40).

dombesz commented on app/controllers/projects/settings/project_custom_fields_controller.rb:
```diff
@@ -0,0 +1,118 @@
+    call = bulk_edit_service.call(action: :enable)
+
+    if call.success?
+      eager_load_project_custom_field_project_mappings # reload mappings
```
Since we can merge the before actions together, we won't have this method anymore. However we could still reload the results by calling:
```suggestion
      @project_custom_field_project_mappings.reload # reload mappings
```

dombesz commented on app/forms/custom_fields/inputs/base/autocomplete/multi_value_input.rb:
```diff
@@ -0,0 +1,62 @@
+  end
+
+  def decorated?
+    raise NotImplementedError
```
Is this intentional?

jjabari-op replied:
Yes as this is a base class

dombesz commented on app/forms/custom_fields/inputs/base/autocomplete/single_value_input.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  def decorated?
+    raise NotImplementedError
```
Is this intentional?

jjabari-op replied:
Yes as this is a base class

dombesz commented on app/models/projects/acts_as_customizable_patches.rb:
```diff
@@ -0,0 +1,248 @@
+    def for_custom_field_accessor(method_symbol)
+      match = /\Acustom_field_(?<id>\d+)=?\z/.match(method_symbol.to_s)
+      if match
+        custom_field = available_custom_fields(global: true).find { |cf| cf.id.to_s == match[:id] }
```
I get an error related to this line on the browser console while loading the overview page.

<img width="1280" alt="image" src="https://github.com/opf/openproject/assets/83396/08096895-db21-4fb1-b5bc-04cc3e4aa527">

dombesz commented on config/initializers/feature_decisions.rb:
```diff
@@ -38,3 +38,5 @@
 #     OpenProject::FeatureDecisions.add :some_flag
 #   end
+
+OpenProject::FeatureDecisions.add :project_attributes
```
Is this in use? Could not find any reference for it.

jjabari-op replied:
Good point! Forgot to remove this!

dombesz commented on app/forms/custom_fields/inputs/text.rb:
```diff
@@ -0,0 +1,38 @@
+    # TODO: rich_text_area not working yet
+    # Uncaught DOMException: Failed to execute 'querySelector' on 'Element': '#project_project[new_custom_field_values_attributes][xyz][value]' is not a valid selector.
+    # --> rich_text_area is not using the configured id, which is not scoped to model via base_config
+    # --> ids with '[' ']' are not valid selectors
```
Is this todo still valid? I don't get the error mentioned here while using the rich text input.

jjabari-op replied:
Already removed :)

dombesz commented on modules/overviews/app/controllers/overviews/overviews_controller.rb:
```diff
@@ -1,14 +1,84 @@
+    end
+
+    def eager_loaded_project_custom_field_sections
+      ProjectCustomFieldSection.all.to_a
```
Calling `to_a` explicitly here too.

jjabari-op replied:
Same reason, to prevent N+1 queries in the view components

dombesz commented on app/components/projects/settings/project_custom_field_sections/show_component.rb:
```diff
@@ -0,0 +1,63 @@
+        end
+
+        def sorted_project_custom_fields
+          @project_custom_fields.sort_by { |pcf| pcf.position_in_custom_field_section }
```
I'm wondering if this is necessary? We already seem to have a method that sorts the custom fields in the acts_as_customizable patch. (There are a few other instances too)

https://github.com/opf/openproject/blob/0ea5c3f36848a92c6311ffab6304702d8fbbc6e7/app/models/projects/acts_as_customizable_patches.rb#L166-L169

jjabari-op replied:
In this component, eager loaded data is accessed, thus it would not be ideal to go for the model method in this case

dombesz commented on app/components/settings/project_custom_field_sections/show_component.rb:
```diff
@@ -0,0 +1,127 @@
+      end
+
+      def move_actions(menu)
+        # TODO: these methods trigger database queries for each section displayed
```
Yes, this will trigger N+1 queries. I have solved a similar issue on the meeting agenda items in this PR #14190 . 
The solution is to pass down the first and last element information to the ShowComponent in order to avoid the N+1 queries.

jjabari-op replied:
Uh nice! Would like to implement here as well? :)

dombesz commented on app/components/settings/project_custom_field_sections/custom_field_row_component.rb:
```diff
@@ -0,0 +1,104 @@
+      end
+
+      def move_actions(menu)
+        # TODO: these methods trigger database queries for each custom field displayed
```
Same here: #14190

dombesz commented on app/components/settings/project_custom_fields/header_component.rb:
```diff
@@ -0,0 +1,41 @@
+      include OpPrimer::ComponentHelpers
+      include OpTurbo::Streamable
+
+      def initialize
```
Is this necessary?

jjabari-op replied:
Some part is already cleaned up, but the `Streamable` module is actually required in this case

dombesz commented on app/controllers/admin/settings/project_custom_field_sections_controller.rb:
```diff
@@ -0,0 +1,102 @@
+    end
+
+    def move
+      @project_custom_field_section.move_to = params[:move_to]&.to_sym
```
It's not mission critical, but we should leave a todo here to handle errors in this controller.
Probably we also need to move these actions into services in the future.

dombesz commented on app/controllers/admin/settings/project_custom_fields_controller.rb:
```diff
@@ -0,0 +1,130 @@
+
+    def drop
+      # prototyopical implementation
+      # needs refactoring via update service
```
I agree.

dombesz commented on app/controllers/custom_fields_controller.rb:
```diff
@@ -36,7 +37,12 @@ class CustomFieldsController < ApplicationController
     # loading wp cfs exclicity to allow for eager loading
-    @custom_fields_by_type = CustomField.all.where.not(type: 'WorkPackageCustomField').group_by { |f| f.class.name }
+    @custom_fields_by_type = CustomField.all
+      .where.not(type: 'WorkPackageCustomField')
+      # ProjectCustomFields now managed in a different UI
+      .tap { |query| query.where.not(type: 'ProjectCustomField') }
```
Just a nitpick, type can also take an array of strings
```suggestion
      .where.not(type: ['WorkPackageCustomField', 'ProjectCustomField'])
```

dombesz reviewed (commented):
Massive job Jonas, congrats!👏
I have left a few comments, admittedly I didn't spent too much time on the UI elements, but they seem pretty straightforward. I tried to comment all my observations just for the record, so if you think some of them should be addressed separately, that is also fine with me.
I haven't go through the whole PR, still have to double check the migrations and check the specs.

jjabari-op committed:
changed n/a text

jjabari-op committed:
adde required label for global project attributes overview

jjabari-op committed:
added parent link for mobile views

dombesz committed:
Update project details widget deprecation message.

jjabari-op committed:
respect visible setting and enriched specs

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
revert temporary change

jjabari-op committed:
fixed custom value specs

jjabari-op committed:
fixed dialog specs

jjabari-op committed:
fixed eslint issues

jjabari-op committed:
fixing admin menu specs

jjabari-op committed:
trying to resolve issues only seen in CI

jjabari-op committed:
remove project custom field specs from other custom field feature specs as they have been moved

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
fix CI issues due to render error in sidebar when no section was assigned

dombesz commented on modules/xls_export/spec/models/xls_export/project/exporter/xls_integration_spec.rb:
```diff
@@ -33,7 +33,8 @@
     before do
-      project.update(description: "This is an <p>html</p> description.")
+      project.description = "This is an <p>html</p> description."
+      project.save!(validate: false)
```
```suggestion
      project.update_column(:description, "This is an <p>html</p> description.")
```

dombesz commented on modules/xls_export/spec/models/xls_export/project/exporter/xls_integration_spec.rb:
```diff
@@ -55,27 +56,34 @@
+
+        # The column for the project-level-disabled custom field is blank
+        expect(sheet.row(1)[header.index(not_used_string_cf.name)]).to be_nil
+        # TODO: CSV export renders "" instead of nil, why does XLS export render nil?
```
This comment seems to be contradictory.

dombesz commented on spec/features/admin/project_custom_fields/create_spec.rb:
```diff
@@ -0,0 +1,106 @@
+RSpec.describe 'Create project custom fields', :js do
+  include_context 'with seeded project custom fields'
+
+  context 'with unsufficient permissions' do
```
Just a typo:
```suggestion
  context 'with insufficient permissions' do
```

jjabari-op committed:
adjusted copy service to not activate project custom fields with default values which are disabled in source

jjabari-op committed:
added specs for invisible custom fields for project creation

jjabari-op committed:
fixed and finetuned project custom field activation for create, update and copy contexts

jjabari-op committed:
adjusted custom value specs

dombesz commented on spec/features/admin/project_custom_fields/create_spec.rb:
```diff
@@ -0,0 +1,106 @@
+    it 'prevents creating a new project custom field with an empty name' do
+      click_on('Save')
+
+      # no server side validation shown, html5 validation is used
```
```suggestion
      expect(page).to have_field 'custom_field_name', validation_message: 'Please fill in this field.'
```

dombesz commented on spec/features/admin/project_custom_fields/edit_spec.rb:
```diff
@@ -0,0 +1,95 @@
+RSpec.describe 'Edit project custom fields', :js do
+  include_context 'with seeded project custom fields'
+
+  context 'with unsufficient permissions' do
```
```suggestion
  context 'with insufficient permissions' do
```

dombesz commented on spec/features/admin/project_custom_fields/edit_spec.rb:
```diff
@@ -0,0 +1,95 @@
+      fill_in('custom_field_name', with: '')
+      click_on('Save')
+
+      # no server side validation shown, html5 validation is used
```
```suggestion
     expect(page).to have_field 'custom_field_name', validation_message: 'Please fill in this field.'
```

dombesz commented on spec/features/admin/project_custom_fields/list_spec.rb:
```diff
@@ -0,0 +1,276 @@
+
+  # helper methods:
+
+  def within_project_custom_field_section_container(section, &block)
```
I think it would be useful to have a support class for the list, something like `Pages::ProjectCustomFieldListPage`. These classes are usually inherited from the `support/pages/page.rb` and they can hold all the logic that is required to navigate around the page.

dombesz commented on spec/features/admin/project_custom_fields/list_spec.rb:
```diff
@@ -0,0 +1,276 @@
+      expect(containers[3].text).to include(section_for_multi_select_fields.name)
+    end
+
+    describe 'managing project custom fields' do
```
I would also add some tests for the "Move" action buttons.

jjabari-op replied:
They are already implemented ;)

dombesz commented on spec/features/projects/copy_spec.rb:
```diff
@@ -140,6 +152,209 @@
+      end
+
+      context 'with non-admin user' do
+        # TBD: Not sure if this is the desired behavior, but would be a bit tricky to change
```
From my point of view, this is the logical behaviour.

dombesz commented on spec/features/projects/copy_spec.rb:
```diff
@@ -140,6 +152,209 @@
+        expect(copied_project.custom_value_for(optional_project_custom_field).typed_value).to eq('')
+      end
+
+      # TBD: Is this intended from a conceptial point of view?
```
Best would be to raise this question to product.

dombesz commented on spec/features/projects/create_spec.rb:
```diff
@@ -133,5 +138,256 @@
+      context 'with a non-admin user' do
+        current_user { create(:user, global_permissions: %i[add_project]) }
+
+        it 'does not show invisible fields in the form and thus not activates the invisible field' do
```
If we would copy the hidden fields, it would possibly expose them to the non-admin user? (Given the non-admin user becomes an admin in the new project.)

jjabari-op replied:
I don't think so as the non-visible fields are only exposed to global admins, not project admins

dombesz commented on spec/permissions/manage_project_custom_values_spec.rb:
```diff
@@ -0,0 +1,50 @@
+  # render sidebar on project overview page with view_project permission
+  # TODO: prevents calling overviews/overviews#project_custom_fields_sidebar when not having the permission view_project (FAILED - 1)
+  it 'prevents calling overviews/overviews#project_custom_fields_sidebar when not having the permission view_project' do
+    pending 'spec failing, reason unknown'
```
Just putting a comment for the pending spec.

dombesz commented on spec/models/projects/exporter/csv_integration_spec.rb:
```diff
@@ -60,29 +60,41 @@
 
     context 'when ee enabled', with_ee: %i[custom_fields_in_projects_list] do
-      it 'renders all those columns' do
+      it 'renders all globally available project custom fields in the header' do
```
Shall we test that hidden fields are excluded from the export?

jjabari-op replied:
Good catch! Have extended specs accordingly

dombesz commented on spec/services/project_custom_field_project_mappings/toggle_service_spec.rb:
```diff
@@ -0,0 +1,217 @@
+             })
+    end
+
+    it 'toggles visible, non-required fields' do
```
These tests could be extracted out into shared_examples.

dombesz commented on spec/services/project_custom_field_project_mappings/toggle_service_spec.rb:
```diff
@@ -0,0 +1,217 @@
+             })
+    end
+
+    it 'toggles visible, non-required fields' do
```
```suggestion
    it 'does not toggle visible, non-required fields' do
```

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
Added uniqueness validation as suggested by @dombesz

jjabari-op committed:
Cleanup as suggested by @dombesz

jjabari-op committed:
Minor cleanups

jjabari-op committed:
Rename service as suggested by @dombesz

jjabari-op committed:
Refactored controller as suggested by @dombesz

jjabari-op committed:
Cleanup as suggested by @dombesz

jjabari-op committed:
Refactoring as suggested by @dombesz

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

dombesz commented on spec/features/projects/project_custom_fields/overview_page/dialog/inputs_spec.rb:
```diff
@@ -0,0 +1,649 @@
+      describe 'with boolean CF' do
+        let(:custom_field) { boolean_project_custom_field }
+        let(:default_value) { false }
+        let(:expected_blank_value) { false }
```
These values does not seem to be used in the `a custom field checkbox` shared examples.

oliverguenther committed:
Fix trackBy function receiving item as first argument

dombesz committed:
Respect closed version setting in multi select version fields

dombesz commented on spec/features/projects/project_custom_fields/overview_page/dialog/update_spec.rb:
```diff
@@ -0,0 +1,656 @@
+      describe 'with boolean CF' do
+        let(:custom_field) { boolean_project_custom_field }
+        let(:field) { FormFields::Primerized::InputField.new(custom_field) }
+        let(:expected_initial_value) { true }
```
This variable is not being used in the expectation.

dombesz commented on spec/features/projects/project_custom_fields/overview_page/dialog/validation_spec.rb:
```diff
@@ -0,0 +1,520 @@
+          dialog.submit
+
+          field.expect_error(I18n.t('activerecord.errors.messages.invalid'))
+        end
```
Do you think these test are necessary? The validation functionality is already tested in the custom field unit tests. The only test we need is to ensure that the error messages are shown.

dombesz commented on spec/features/projects/project_custom_fields/overview_page/dialog/validation_spec.rb:
```diff
@@ -0,0 +1,520 @@
+
+        it_behaves_like 'a custom field input'
+
+        it 'shows an error if the value is too long' do
```
Same for these.

dombesz commented on spec/features/projects/project_custom_fields/overview_page/sidebar_spec.rb:
```diff
@@ -0,0 +1,860 @@
+    end
+  end
+
+  describe 'with correct order' do
```
With adjusting the specs from the "with correct scoping" we can also ensure the order is correct, and we can remove these specs.

dombesz commented on spec/features/projects/project_custom_fields/settings/mapping_spec.rb:
```diff
@@ -0,0 +1,448 @@
+    within("[data-qa-selector='project-custom-field-section-#{section.id}']", &)
+  end
+
+  def within_custom_field_container(custom_field, &)
```
Just as a future improvement we could create a `ProjectCustomFields` support view class.

dombesz commented on frontend/src/stimulus/controllers/dynamic/projects/settings/project-custom-fields-mapping-filter.controller.ts:
```diff
@@ -0,0 +1,95 @@
+    this.showBulkActionContainers();
+
+    this.searchItemTargets.forEach((item) => {
+      (item as HTMLElement).style.display = 'block';
```
I was wondering if this could cause problems later on, in case the `display` is going to be changed to another value in the styling. Perhaps it would be better to add and remove a `.hidden` class?

jjabari-op replied:
Good catch!

dombesz commented on app/views/admin/settings/projects_settings/show.html.erb:
```diff
@@ -60,6 +60,7 @@ See COPYRIGHT and LICENSE files for more details.
         </div>
       <% end %>
 
+
```
Just an extra line here.

dombesz commented on app/components/settings/project_custom_field_sections/show_component.rb:
```diff
@@ -0,0 +1,127 @@
+        super
+
+        @project_custom_field_section = project_custom_field_section
+        @project_custom_fields = project_custom_field_section.custom_fields.reorder(position_in_custom_field_section: :asc)
```
The reorder can be removed, since it is being added to the original query, this way will not trigger an N+1 query.

dombesz commented on app/services/projects/copy_service.rb:
```diff
@@ -80,6 +80,42 @@ def before_perform(params, service_call)
+      custom_fields_to_enable = custom_fields_activated_in_source - custom_fields_activated_in_target
+      call.result.project_custom_fields << ProjectCustomField.where(id: custom_fields_to_enable)
+    end
+
```
This can be simplified, we need everything activated from the source project in the target project and nothing else.

```suggestion
    def after_perform(call)
      copy_activated_custom_fields(call)

      super
    end

    def copy_activated_custom_fields(call)
       call.result.project_custom_field_ids = source.project_custom_field_ids
    end
```

dombesz commented on db/migrate/20240208100316_enable_required_project_custom_fields_in_all_projects.rb:
```diff
@@ -0,0 +1,17 @@
+          ProjectCustomFieldProjectMapping.create!(project_id: project.id, custom_field_id: pcf.id)
+        end
+      end
+    end
```
This code is not really efficient, it loads up all the required fields and all the projects into objects which can be costly on a large # of projects or custom fields. Additionally it will run N+1 insert queries which can take a long time. I suggest using just the ids instead of objects and the `insert_all!` method to avoid the N+1 inserts.

```suggestion
    required_custom_field_ids = ProjectCustomField.required.ids

    # Gather the custom_field_ids for every project, then add a new mapping
    # of {project_id:, custom_field_id:} for every project that does not have
    # the required required_custom_field_ids activated.
    missing_custom_field_attributes =
      Project
        .includes(:project_custom_field_project_mappings)
        .pluck(:id, 'project_custom_field_project_mappings.custom_field_id')
        .group_by(&:first)
        .transform_values { |values| values.map(&:last) }
        .reduce([]) do |acc, (project_id, custom_field_ids)|

        missing_custom_field_ids = required_custom_field_ids - custom_field_ids

        acc + missing_custom_field_ids.map do |custom_field_id|
          { project_id: , custom_field_id: }
        end
      end

    ProjectCustomFieldProjectMapping.insert_all!(missing_custom_field_attributes)
```

This code will load the ids only, and inserts all the missing mappings in one instert query.

jjabari-op committed:
Refactored prototypical implementation as suggested by @dombesz

jjabari-op committed:
Refactored to a service based approach as suggested by @dombesz

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

dombesz commented on spec/models/projects/customizable_spec.rb:
```diff
@@ -27,92 +27,407 @@
+      context 'when custom fields are mapped to this project' do
+        before do
+          project.project_custom_fields << [text_custom_field, bool_custom_field]
+          project.reload # TODO: why is this necessary?
```
I believe reload is not required anymore.

dombesz commented on spec/models/projects/customizable_spec.rb:
```diff
@@ -27,92 +27,407 @@
+      end
+    end
+
+    it 'does re-enable fields with new value which have been disabled in the past' do
```
Is this the usecase to ensure that updates coming from the api will accept the disabled custom fields too? I'm wondering if we should raise a validation error instead?

jjabari-op replied:
I figured that's the way we enable the implicit control over the activation state via the API without further API adjustments.

dombesz reviewed (commented):
Finished the second round of review, I will have to review the latest changes to wrap up the review completely.

jjabari-op committed:
fixing last CI errors, mostly related to adjusted behaviour of the angular autocomplete input

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

dombesz commented on app/components/settings/project_custom_fields/header_component.html.erb:
```diff
@@ -0,0 +1,36 @@
+          action_buttons_container.with_column(mr: 2) do
+            render(Primer::Beta::Button.new(
+              tag: :a,
+              href: new_admin_settings_project_custom_field_path(type: "ProjectCustomField"),
```
I think we should disable this button if there are no `@project_custom_field_sections` are present and also validate the presence of the `ProjectCustomField#custom_field_section_id` field. Currently it is possible to create a project custom field without a section id, which is causing an error while rendering the overview page.
<details><summary>Demo of bug</summary>
<p>

![attribute missing](https://github.com/opf/openproject/assets/83396/0ba4c705-4bdf-40d8-8ecd-9c081a94156d)

</p>
</details>

jjabari-op committed:
Update modules/xls_export/spec/models/xls_export/project/exporter/xls_integration_spec.rb

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

jjabari-op committed:
Update spec/features/admin/project_custom_fields/create_spec.rb

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

jjabari-op committed:
Update spec/features/admin/project_custom_fields/create_spec.rb

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

jjabari-op committed:
Update spec/features/admin/project_custom_fields/edit_spec.rb

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

jjabari-op committed:
Update spec/features/admin/project_custom_fields/edit_spec.rb

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

jjabari-op committed:
checking for correct handling of invisible fields in export specs as suggested by @dombesz

jjabari-op committed:
cleanup resolved comment

jjabari-op committed:
Update spec/services/project_custom_field_project_mappings/toggle_service_spec.rb

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

dombesz committed:
Rename project custom field project mappings contract validation methods.

oliverguenther committed:
Fix invitation spec

invited users don't have an href so we cant track them on it

dombesz committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

dombesz committed:
Avoid N+1 queries on the ProjectCustomFieldSections

Additionally moving the custom_fields relation from the CustomFieldSection
to the ProjectCustomFieldSection. This is required, because we have to specify the
custom_fields relation to include ProjectCustomField objects only.
Otherwise the nested include query in the ProjectCustomFieldsController does not work.

jjabari-op committed:
cleanup

jjabari-op committed:
Fixed model relation definitions

jjabari-op committed:
Fixed specs

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op committed:
Fixed customizable spec

jjabari-op committed:
Update app/services/projects/copy_service.rb

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

jjabari-op committed:
Update db/migrate/20240208100316_enable_required_project_custom_fields_in_all_projects.rb

Co-authored-by: Dombi Attila <83396+dombesz@users.noreply.github.com>

jjabari-op committed:
validate presence of custom_field_section as suggested by @dombesz

jjabari-op committed:
Spec adjustments as suggested by @dombesz

jjabari-op committed:
Adjusted filter approach as suggested by @dombesz

dombesz committed:
Move ProjectCustomField section validation to the object instead of the field.

jjabari-op committed:
quick fix failing spec due to changed error message

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

dombesz committed:
Display error messages for missing custom field section on creating project custom fields.

dombesz committed:
Remove N+1 queries from CustomFieldSection rendering

dombesz commented on app/controllers/concerns/projects/settings/project_custom_fields/component_streams.rb:
```diff
@@ -0,0 +1,69 @@
+            )
+          end
+
+          def update_custom_field_row_via_turbo_stream(
```
This method does not seem to be used.

dombesz committed:
Remove N+1 queries from the ProjectCustomFields rendering inside the custom field sections

jjabari-op committed:
fixed spec due to changed validation approach

jjabari-op committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

dombesz committed:
Simplify the Project settings project custom fields page.

dombesz committed:
Add visible scope

dombesz committed:
Simplify the project acts as customizable patch, load all the field data in one query and sort it via sql instead of ruby.

dombesz committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

dombesz committed:
Do not pass down the project.project_custom_field_project_mappings to the CustomFieldRowComponent but call it from the project object.

This is safe to do, because we are calling the project_custom_field_project_mappings
on the same project object multiple times and this won't result in an N+1 query.

dombesz committed:
Fix typo

dombesz committed:
Use translation to count projects on admin project custom field row

dombesz committed:
Fix comment typo

dombesz committed:
Use hidden by filter class on the batch actions too

dombesz reviewed (approved):
Well done @jjabari-op 🎉

dombesz committed:
Merge branch 'dev' into epic/49688-custom-set-of-project-attributes-grouped-in-sections

jjabari-op commented:
Thank you for the review, improvements and merge @dombesz

Eric-Guo commented:
```sql
Migrating to CreateCustomFieldSections (20231123111357)
== 20231123111357 CreateCustomFieldSections: migrating ========================
-- create_table(:custom_field_sections)
  TRANSACTION (0.1ms)  BEGIN
   (4.7ms)  CREATE TABLE "custom_field_sections" ("id" bigserial primary key, "position" integer, "name" character varying, "type" character varying, "created_at" timestamptz(6) NOT NULL, "updated_at" timestamptz(6) NOT NULL)
   -> 0.0051s
-- add_reference(:custom_fields, :custom_field_section)
   (0.3ms)  ALTER TABLE "custom_fields" ADD "custom_field_section_id" bigint
   (1.3ms)  CREATE INDEX "index_custom_fields_on_custom_field_section_id" ON "custom_fields" ("custom_field_section_id")
   -> 0.0046s
-- add_column(:custom_fields, :position_in_custom_field_section, :integer, {:null=>true})
   (0.2ms)  ALTER TABLE "custom_fields" ADD "position_in_custom_field_section" integer
   -> 0.0006s
  CustomFieldSection Load (0.5ms)  SELECT "custom_field_sections".* FROM "custom_field_sections" WHERE "custom_field_sections"."type" = $1 AND ("custom_field_sections"."position" IS NOT NULL) ORDER BY "custom_field_sections"."position" DESC LIMIT $2  [["type", "ProjectCustomFieldSection"], ["LIMIT", 1]]
  ProjectCustomFieldSection Create (0.6ms)  INSERT INTO "custom_field_sections" ("position", "name", "type", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["position", 1], ["name", "Project attributes"], ["type", "ProjectCustomFieldSection"], ["created_at", "2024-03-25 14:53:54.495725"], ["updated_at", "2024-03-25 14:53:54.495725"]]
  ProjectCustomField Load (0.7ms)  SELECT "custom_fields".* FROM "custom_fields" WHERE "custom_fields"."type" = $1 ORDER BY "custom_fields"."id" ASC LIMIT $2  [["type", "ProjectCustomField"], ["LIMIT", 1000]]
  CustomOption Load (0.6ms)  SELECT "custom_options".* FROM "custom_options" WHERE "custom_options"."custom_field_id" = $1 ORDER BY "custom_options"."position" ASC  [["custom_field_id", 13]]
  CustomField Pluck (0.2ms)  SELECT "custom_fields"."name" FROM "custom_fields" WHERE "custom_fields"."type" = $1 AND "custom_fields"."id" != $2  [["type", "ProjectCustomField"], ["id", 13]]
  ↳ app/models/custom_field.rb:57:in `uniqueness_of_name_with_scope'
  CustomOption Pluck (0.2ms)  SELECT "custom_options"."id" FROM "custom_options" WHERE "custom_options"."custom_field_id" = $1 AND "custom_options"."default_value" = $2 ORDER BY "custom_options"."position" ASC  [["custom_field_id", 13], ["default_value", true]]
  ↳ app/models/custom_field.rb:83:in `default_value'
  ProjectCustomField Load (1.3ms)  SELECT "custom_fields".* FROM "custom_fields" WHERE "custom_fields"."custom_field_section_id" = $1 AND ("custom_fields"."position_in_custom_field_section" IS NOT NULL) ORDER BY "position_in_custom_field_section" DESC LIMIT $2  [["custom_field_section_id", 1], ["LIMIT", 1]]
  ProjectCustomField Update (0.8ms)  UPDATE "custom_fields" SET "updated_at" = $1 WHERE "custom_fields"."id" = $2  [["updated_at", "2024-03-25 14:53:54.548179"], ["id", 13]]
  ProjectCustomField Load (0.4ms)  SELECT "custom_fields".* FROM "custom_fields" WHERE "custom_fields"."custom_field_section_id" = $1 AND ("custom_fields"."position_in_custom_field_section" IS NOT NULL) ORDER BY "position_in_custom_field_section" DESC LIMIT $2  [["custom_field_section_id", 1], ["LIMIT", 1]]
  ProjectCustomField Count (0.3ms)  SELECT COUNT(*) FROM "custom_fields" WHERE "custom_fields"."custom_field_section_id" = $1 AND ("custom_fields"."position_in_custom_field_section" = 1)  [["custom_field_section_id", 1]]
== 20231123111357 CreateCustomFieldSections: migrated (0.0813s) ===============

  ActiveRecord::SchemaMigration Create (0.2ms)  INSERT INTO "schema_migrations" ("version") VALUES ('20231123111357') RETURNING "version"
  TRANSACTION (0.7ms)  COMMIT
```

Nice work, but I found when first doing the CreateCustomFieldSections migrate, existing `ProjectCustomField` (in this case, custom_fields.id==13) the columns custom_field_section_id is not filled, maybe due to custom_field_section is just added.

So I re-run below code and got the custom_field_section_id updated.

```ruby
[2] pry(main)> ProjectCustomField.first
  ProjectCustomField Load (0.6ms)  SELECT "custom_fields".* FROM "custom_fields" WHERE "custom_fields"."type" = $1 ORDER BY "custom_fields"."id" ASC LIMIT $2  [["type", "ProjectCustomField"], ["LIMIT", 1]]
=> #<ProjectCustomField:0x0000000134610960
 id: 13,
 type: "ProjectCustomField",
 field_format: "list",
 regexp: "",
 min_length: 0,
 max_length: 0,
 is_required: false,
 is_for_all: false,
 is_filter: false,
 position: 1,
 searchable: true,
 editable: true,
 visible: true,
 multi_value: false,
 default_value: nil,
 name: "测试项目字段",
 created_at: Thu, 25 Jan 2024 13:53:11.813609000 CST +08:00,
 updated_at: Mon, 25 Mar 2024 22:53:54.548179000 CST +08:00,
 content_right_to_left: false,
 allow_non_open_versions: false,
 custom_field_section_id: nil,
 position_in_custom_field_section: nil>
[3] pry(main)> section = ProjectCustomFieldSection.create!(    
[3] pry(main)*   name: "Project attributes 2"      
[3] pry(main)* )      
  TRANSACTION (0.1ms)  BEGIN
  CustomFieldSection Load (1.2ms)  SELECT "custom_field_sections".* FROM "custom_field_sections" WHERE "custom_field_sections"."type" = $1 AND ("custom_field_sections"."position" IS NOT NULL) ORDER BY "custom_field_sections"."position" DESC LIMIT $2  [["type", "ProjectCustomFieldSection"], ["LIMIT", 1]]
  ProjectCustomFieldSection Create (0.5ms)  INSERT INTO "custom_field_sections" ("position", "name", "type", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["position", 2], ["name", "Project attributes 2"], ["type", "ProjectCustomFieldSection"], ["created_at", "2024-03-25 14:56:14.169673"], ["updated_at", "2024-03-25 14:56:14.169673"]]
  TRANSACTION (0.9ms)  COMMIT
=> #<ProjectCustomFieldSection:0x0000000169bb6b78
 id: 2,
 position: 2,
 name: "Project attributes 2",
 type: "ProjectCustomFieldSection",
 created_at: Mon, 25 Mar 2024 22:56:14.169673000 CST +08:00,
 updated_at: Mon, 25 Mar 2024 22:56:14.169673000 CST +08:00>
[4] pry(main)> ProjectCustomField.find_each do |project_custom_field|    
[4] pry(main)*   project_custom_field.update!(custom_field_section_id: section.id)      
[4] pry(main)* end      
  ProjectCustomField Load (0.4ms)  SELECT "custom_fields".* FROM "custom_fields" WHERE "custom_fields"."type" = $1 ORDER BY "custom_fields"."id" ASC LIMIT $2  [["type", "ProjectCustomField"], ["LIMIT", 1000]]
  TRANSACTION (3.6ms)  BEGIN
  CustomOption Load (0.7ms)  SELECT "custom_options".* FROM "custom_options" WHERE "custom_options"."custom_field_id" = $1 ORDER BY "custom_options"."position" ASC  [["custom_field_id", 13]]
  CustomField Pluck (0.3ms)  SELECT "custom_fields"."name" FROM "custom_fields" WHERE "custom_fields"."type" = $1 AND "custom_fields"."id" != $2  [["type", "ProjectCustomField"], ["id", 13]]
  CustomOption Pluck (0.2ms)  SELECT "custom_options"."id" FROM "custom_options" WHERE "custom_options"."custom_field_id" = $1 AND "custom_options"."default_value" = $2 ORDER BY "custom_options"."position" ASC  [["custom_field_id", 13], ["default_value", true]]
  ProjectCustomField Load (0.3ms)  SELECT "custom_fields".* FROM "custom_fields" WHERE "custom_fields"."custom_field_section_id" = $1 AND ("custom_fields"."position_in_custom_field_section" IS NOT NULL) ORDER BY "custom_fields"."position_in_custom_field_section" DESC LIMIT $2  [["custom_field_section_id", 2], ["LIMIT", 1]]
  ProjectCustomField Update (0.4ms)  UPDATE "custom_fields" SET "updated_at" = $1, "custom_field_section_id" = $2, "position_in_custom_field_section" = $3 WHERE "custom_fields"."id" = $4  [["updated_at", "2024-03-25 14:56:29.008077"], ["custom_field_section_id", 2], ["position_in_custom_field_section", 1], ["id", 13]]
  ProjectCustomField Load (0.2ms)  SELECT "custom_fields".* FROM "custom_fields" WHERE "custom_fields"."custom_field_section_id" = $1 AND ("custom_fields"."position_in_custom_field_section" IS NOT NULL) ORDER BY "custom_fields"."position_in_custom_field_section" DESC LIMIT $2  [["custom_field_section_id", 2], ["LIMIT", 1]]
  ProjectCustomField Count (0.2ms)  SELECT COUNT(*) FROM "custom_fields" WHERE "custom_fields"."custom_field_section_id" = $1 AND ("custom_fields"."position_in_custom_field_section" = 1)  [["custom_field_section_id", 2]]
  TRANSACTION (1.3ms)  COMMIT
=> nil
```

I try at TH staging server, it will lead 500 error due to custom_field_section_id is NULL. I try running above code to fix and it works.

dombesz commented:
Thanks for the report @Eric-Guo ! I have filed a bug report on your behalf in our community instance https://community.openproject.org/wp/53728 . We will look into this before the release.
